testtime/testtime.go

```diff
@@ -0,0 +1,213 @@
+		c:          c,
+	}
+	if d <= 0 {
+		defer timer.doFire(timer.currTime)
+	}
+	return timer
+}
+
+// ExpiredC returns the underlying C channel of the timer.
+func (t *TestTimer) ExpiredC() <-chan time.Time {
+	return t.c
+}
+
+// Reset changes the timer to expire after duration d. It returns true if the
+// timer had been active, false if the timer had expired or been stopped.
+//
+// As the test timer does not actually count down, Reset sets the timer's
+// expiration to be the given duration added to the timer's internal current
+// time. This internal time must be advanced manually using Elapse.
+//
+// This simulates the behavior of Timer.Reset() from the time package.
+// See here fore more details: https://pkg.go.dev/time#Timer.Reset
+func (t *TestTimer) Reset(d time.Duration) bool {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	active := t.active
+	t.active = true
+	t.expiration = t.currTime.Add(d)
+	if t.c != nil {
+		// Drain the channel, guaranteeing that a receive after Reset will
+		// block until the timer fires again, and not receive a time value
+		// from the timer firing before the reset occurred.
+		// This complies with the new behavior of Reset as of Go 1.23.
+		// See: https://pkg.go.dev/time#Timer.Reset
+		select {
+		case <-t.c:
+		default:
+		}
+	}
+	// If duration is 0 or negative, ensure timer fires
+	if d <= 0 {
+		defer t.doFire(t.currTime)
+	}
+	return active
+}
+
+// Stop prevents the timer from firing. It returns true if the call stops the
+// timer, false if the timer has already expired or been stopped.
+//
+// This simulates the behavior of Timer.Stop() from the time package.
+// See here for more details: https://pkg.go.dev/time#Timer.Stop
+func (t *TestTimer) Stop() bool {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	wasActive := t.active
+	t.active = false
+	if t.c != nil {
+		// Drain the channel, guaranteeing that a receive after Stop will block
+		// and not receive a time value from the timer firing before the stop
+		// occurred. This complies with the new behavior of Stop as of Go 1.23.
+		// See: https://pkg.go.dev/time#Timer.Stop
+		select {
+		case <-t.c:
+		default:
+		}
+	}
+	return wasActive
+}
+
+// Active returns true if the timer is active, false if the timer has expired
+// or been stopped.
+func (t *TestTimer) Active() bool {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	return t.active
+}
+
+// FireCount returns the number of times the timer has fired.
+func (t *TestTimer) FireCount() int {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	return t.fireCount
+}
+
+// Elapse simulates the current time advancing by the given duration, which
+// potentially causes the timer to fire.
+//
+// The timer will fire if the time after the elapsed duration is after the
+// expiration time and the timer has not yet fired.
+func (t *TestTimer) Elapse(duration time.Duration) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	t.currTime = t.currTime.Add(duration)
+	if !t.currTime.Before(t.expiration) {
+		t.doFire(t.expiration)
```

Comment:
this could use time.Now then? if somebody cares about the time, then they could use Fire?